### PR TITLE
P3A rewards duration

### PR DESCRIPTION
**Pull request by rillian**
Resolves brave/brave-browser#18394

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by rillian** [comment]
I'm stuck on this and would appreciate some help. The idea is to record how long Rewards were on when they're turned off. `RewardsServiceImpl` seems to use two prefs to track this, one for ads and one for autocontribute.

The test I wrote didn't pass because flipping the pref the way the other browser tests do doesn't propagate synchronously back to `RewardsServiceImpl::OnPreferenceChanged` where I inserted the P3A measurement. The logging in the patch suggested they were running after the test finished, so I wrote a `WaitForRewardsEnable` method which spins the message loop. 

That's enough to get the first "rewards are enabled" stage of measurement to validate, but when I turn rewards off again, the AutoContribute disable doesn't propagate synchronously. If I again run the message loop, it DCHECKS disabling before initialisation is complete.

What else could I wait on to make sure the enable/disable signals have run to completion?Suggestions very welcome!

**Comment by goodov** [comment]
I recommend use these things to debug and better understand what exactly is going on and how to properly write a test:
1. `base::debug::StackTrace(<optional frames count>).Print();` in the method you want to catch to better understand what exactly you need to observe and what test helpers you might need to create. I recommend putting it into `AdsServiceImpl::SetEnabled`.
2. `base::RunLoop().Run();` to halt the test at any point, but keep the browser running. This can help rule out some test-running issues (like timeouts or some mocked parts) and see if the browser actually behaves the way you expect it.
3. `--no-sandbox` to receive logs from other processes (if you need this).
4. `LOG(ERROR)` instead of `std::cerr` to see at what line of what file some event has happened. You just can do `LOG(ERROR);` without any message as a quick tool for "where the hell does this whole chain break?".

Re your test. The `prefs_->Set/Get...()` and observers should always be synchronous, because the prefs can be accessed only on main thread. I think the issue you're having is some asynchronous process in `RewardsService` which waits for something and only then modifies a pref value.

**Comment by rillian** [comment]
I continued to have a lot of trouble getting tests to work. These are a bit hacky, just flipping the pref rather than asking the service to enable/disable, but doing so seemed to interfere with other Rewards browser tests. Maybe there's contention for the same ledger process when they're running in parallel.

In any case, this version seems to pass ci, so ready for review again @iefremov, @brave/rewards-client.

**Comment by rillian** [comment]
Review ping @iefremov, @brave/rewards-client

File: components/brave_rewards/browser/rewards_p3a.cc [hunk 1]
```diff
@@ -180,6 +180,63 @@ double CalcWalletBalance(base::flat_map<std::string, double> wallets,
...
+    // Previous timestamp available.
+    if (!rewards_enabled) {
+      // Rewards have been disabled. Record the duration they were on.
+      // Set the threshold at three units so each bin represents the
+      // nominal value as an order-of-magnitude: more than three days
+      // is a week, more than three weeks is a month, and so on.
+      constexpr auto threshold = 3;
+      constexpr auto days_per_week = 7;
+      constexpr auto days_per_month = 30.44;  // average length
+      auto duration = base::Time::Now() - enabled_timestamp;
```
Comment (iefremov) [inline components/brave_rewards/browser/rewards_p3a.cc:211]: personally I'd not overuse `auto`, e.g. in this case it's convenient to see that `duration` is a TimeDelta
there is a styleguide entry about that https://google.github.io/styleguide/cppguide.html#Type_deduction

File: components/brave_rewards/browser/rewards_p3a.h [hunk 2]
```diff
@@ -73,6 +73,19 @@ void RecordNoWalletCreatedForAllMetrics();
...
 void RecordRewardsDisabledForSomeMetrics();
 
+enum class RewardsEnabledDuration {
+  kNever,
+  kStillEnabled,
+  kHours,
+  kDays,
+  kWeeks,
+  kMonths,
+  kLonger,
```
Comment (iefremov) [inline components/brave_rewards/browser/rewards_p3a.h:83]: longer than what? mb there could be a better name

**Comment by iefremov** [comment]
not finished yet, will take another look

File: components/brave_rewards/browser/rewards_p3a.h (hunk 2 above)
Comment (rillian) [reply components/brave_rewards/browser/rewards_p3a.h:83]: `kMoreThanMonths`? `kSeasons`?

Comment (emerick) [reply components/brave_rewards/browser/rewards_p3a.h:83]: `kQuarters`?

Comment (rillian) [reply components/brave_rewards/browser/rewards_p3a.h:83]: Appropriately corporate for "analytics" :)

File: components/brave_rewards/browser/rewards_p3a.cc [hunk 3]
```diff
@@ -180,6 +180,63 @@ double CalcWalletBalance(base::flat_map<std::string, double> wallets,
...
+    if (!rewards_enabled) {
+      // Rewards have been disabled. Record the duration they were on.
+      // Set the threshold at three units so each bin represents the
+      // nominal value as an order-of-magnitude: more than three days
+      // is a week, more than three weeks is a month, and so on.
+      constexpr int threshold = 3;
+      constexpr int days_per_week = 7;
+      constexpr double days_per_month = 30.44;  // average length
+      base::TimeDelta duration = base::Time::Now() - enabled_timestamp;
+      DLOG(INFO) << "Rewards disabled after " << duration;
```
Comment (iefremov) [inline components/brave_rewards/browser/rewards_p3a.cc:212]: I'd rather say VLOG, afaik most people don't build the debug browser

File: components/brave_rewards/browser/test/rewards_p3a_browsertest.cc [hunk 4]
```diff
@@ -140,6 +175,86 @@ IN_PROC_BROWSER_TEST_F(RewardsP3ABrowserTest, RewardsDisabled) {
...
+                                       RewardsEnabledDuration::kMonths, 1);
+
+  // Mock turning rewards on for our longest measured value.
+  prefs->SetBoolean("brave.brave_ads.enabled", true);
+  auto long_ago = base::Time::Now() - base::Days(128);
+  LOG(INFO) << "Backdating timestamp to " << long_ago;
+  prefs->SetTime(brave_rewards::prefs::kEnabledTimestamp, long_ago);
+
+  // Mock turning rewards off.
+  prefs->SetBoolean("brave.brave_ads.enabled", false);
```
Comment (iefremov) [inline components/brave_rewards/browser/test/rewards_p3a_browsertest.cc:255]: why not `prefs::kEnabled` everywhere?

File: components/brave_rewards/browser/test/rewards_p3a_browsertest.cc [hunk 5]
```diff
@@ -140,6 +175,86 @@ IN_PROC_BROWSER_TEST_F(RewardsP3ABrowserTest, RewardsDisabled) {
...
+  // is made by the rewards service preference observer.
+  prefs->SetBoolean("brave.brave_ads.enabled", false);
+  histogram_tester_->ExpectBucketCount("Brave.Rewards.EnabledDuration",
+                                       RewardsEnabledDuration::kHours, 1);
+
+  // Mock turning rewards back on.
+  prefs->SetBoolean("brave.brave_ads.enabled", true);
+  // Adjust the stored timestamp to measure a longer duration.
+  auto earlier = base::Time::Now() - base::Minutes(90);
+  LOG(INFO) << "Backdating timestamp to " << earlier;
```
Comment (iefremov) [inline components/brave_rewards/browser/test/rewards_p3a_browsertest.cc:207]: VLOG ?

**Review by iefremov — APPROVED** [review]
LGTM with small nits

**Comment by rillian** [comment]
Thanks for the review. Nits addressed in follow-up commits.

File: components/brave_rewards/browser/rewards_p3a.cc (hunk 3 above)
Comment (rillian) [reply components/brave_rewards/browser/rewards_p3a.cc]: Ok, I've changed to `VLOG(1)` in 9d97e1cc3ddc2a343040036205bef9f372355cc7.

NB the default developer component build sets `DCHECK_ALWAYS_ON` which I believe [enables DLOG](https://github.com/chromium/chromium/blob/96.0.4664.93/base/logging.h#L526-L551) as well.

File: components/brave_rewards/browser/test/rewards_p3a_browsertest.cc (hunk 5 above)
Comment (rillian) [reply components/brave_rewards/browser/test/rewards_p3a_browsertest.cc]: I think `LOG` is better for tests.  The output is normally captured by the test harness, so it doesn't clutter normal runs, but is then shown as part of the report when a test fails, without having to figure out how to invoke the test with `-v=1` or whatever.

Nevertheless, replaced in  9d97e1cc3ddc2a343040036205bef9f372355cc7.

File: components/brave_rewards/browser/test/rewards_p3a_browsertest.cc (hunk 4 above)
Comment (rillian) [reply components/brave_rewards/browser/test/rewards_p3a_browsertest.cc]: Originally there was only one reference, and I had trouble finding the correct header. Fixed in 55ec9c0687bfb68742af6498bc76f0243aa06b1e.

**Comment by rillian** [comment]
@brave/rewards-client review ping? Looks like this change needs your approval as well.

File: components/brave_rewards/browser/rewards_p3a.h [hunk 6]
```diff
@@ -73,6 +73,19 @@ void RecordNoWalletCreatedForAllMetrics();
 
 void RecordRewardsDisabledForSomeMetrics();
 
+enum class RewardsEnabledDuration {
+  kNever,
+  kStillEnabled,
```
Comment (emerick) [inline components/brave_rewards/browser/rewards_p3a.h:78]: This value seems a bit out of place in this enumeration, as it doesn't represent a duration per se. Does this mean that it's of unknown duration? If so, maybe its name should reflect that.

File: components/brave_rewards/browser/rewards_p3a.cc [hunk 7]
```diff
@@ -180,6 +180,63 @@ double CalcWalletBalance(base::flat_map<std::string, double> wallets,
   return balance_minus_grant;
 }
 
+void RecordRewardsEnabledDuration(PrefService* prefs, bool rewards_enabled) {
+  base::Time enabled_timestamp = prefs->GetTime(prefs::kEnabledTimestamp);
+  auto value = RewardsEnabledDuration::kNever;
```
[the remark this answers is not in the document]
Comment (rillian) [reply components/brave_rewards/browser/rewards_p3a.cc:185]: Ok.

File: components/brave_rewards/browser/rewards_p3a.h (hunk 6 above)
Comment (rillian) [reply components/brave_rewards/browser/rewards_p3a.h:78]: There are two values representing corner cases where we can't measure a duration. Both are specified as answers in the original P3A question:

- Rewards have never been enabled
- Rewards are currently (still) enabled

We're asking how long rewards were enabled for after someone turns them off. So it's not that the duration is unknown in these states, it's that the question doesn't apply.

Are you asking for something like `RewardsEnabledDuration::kUnknownBecauseNotDisabled`? I was trying to keep the length similar to the duration values, while keying on the 'enabled' part of the type name.

Comment (emerick) [reply components/brave_rewards/browser/rewards_p3a.h:78]: Gotcha, I think that's fine then.

**Review by emerick — APPROVED** [review]

**Pull request merged**